### PR TITLE
Swerve Drives

### PR DESCRIPTION
**Pull request by fizzlebert**
This is the current state of swerve drive code.

### CODE REVIEW

File: robot.py [hunk 1]
```diff
@@ -1,21 +1,147 @@
...
+            reverse_drive_direction=True,
+        )
+        self.module_b = SwerveModule(  # bottom left module now front right
+            "b",
+            steer_talon=ctre.TalonSRX(3),
+            drive_talon=ctre.TalonSRX(4),
+            x_pos=-x_dist,
+            y_pos=y_dist,
+            drive_free_speed=Robot.module_drive_free_speed,
+            reverse_drive_encoder=True,
```
Comment (auscompgeek) [inline robot.py]: Is this correct? Are you telling me that module b wasn't assembled correctly?

File: robot.py [hunk 2]
```diff
@@ -1,21 +1,147 @@
...
+from networktables import NetworkTables
+
+from pyswervedrive.swervechassis import SwerveChassis
+from pyswervedrive.swervemodule import SwerveModule
+from utilities.functions import rescale_js, constrain_angle
+from utilities.navx import NavX
 
 
 class Robot(magicbot.MagicRobot):
+    module_drive_free_speed: float = 94000.0  # encoder ticks / 100 ms
```
Comment (auscompgeek) [inline robot.py]: Why is this here?

File: robot.py [hunk 3]
```diff
@@ -1,21 +1,147 @@
...
+    #  TODO do the maths to verify that this is correct
+    offset_rotation_rate = 40
```
Comment (auscompgeek) [inline robot.py]: What is this for?

File: robot.py [hunk 4]
```diff
@@ -30,7 +28,7 @@ def createObjects(self):
             drive_talon=ctre.TalonSRX(2),
             x_pos=x_dist,
             y_pos=y_dist,
-            drive_free_speed=Robot.module_drive_free_speed,
+            drive_free_speed=SwerveModule.DRIVE_FREE_SPEED,
```
Comment (auscompgeek) [inline robot.py]: Why is this being passed in here?

Comment (fizzlebert) [reply robot.py]: Actually it should be a default argument no being passed in, right?

Comment (fizzlebert) [reply robot.py]: Or not even as an argument

File: robot.py [hunk 5]
```diff
@@ -1,21 +1,141 @@
...
+        # a + + b + - c - - d - +
+        x_dist = 0.2165
+        y_dist = 0.2625
+        self.module_a = SwerveModule(  # top right module now back left
+            "a",
+            steer_talon=ctre.TalonSRX(1),
+            drive_talon=ctre.TalonSRX(2),
+            x_pos=x_dist,
+            y_pos=y_dist,
+            reverse_steer_encoder=True,
```
Comment (auscompgeek) [inline robot.py]: Since we know the encoder phase is reversed on all the swerve modules (in fact this is trivially determined from the CAD), shouldn't this be the default?

File: robot.py [hunk 6]
```diff
@@ -1,21 +1,141 @@
...
+    chassis: SwerveChassis
+
     def createObjects(self):
         """Create motors and stuff here."""
-        pass
+
+        # a + + b + - c - - d - +
+        x_dist = 0.2165
+        y_dist = 0.2625
+        self.module_a = SwerveModule(  # top right module now back left
```
Comment (auscompgeek) [inline robot.py]: These comments make no sense.

**Review by auscompgeek — APPROVED** [review]
Gonna merge this for now, pending an uplift of last year's cleanup.
